extensions/big-data-cluster/src/bigDataCluster/tree/controllerTreeDataProvider.ts

```diff
@@ -0,0 +1,147 @@
+	}
+
+	public addController(
+		url: string,
+		username: string,
+		password: string,
+		rememberPassword: boolean,
+		masterInstance?: IEndPoint
+	): void {
+		this.root.addControllerNode(url, username, password, rememberPassword, masterInstance);
+		this.notifyNodeChanged();
+	}
+
+	public deleteController(url: string, username: string): ControllerNode {
+		let deleted = this.root.deleteControllerNode(url, username);
+		if (deleted) {
+			this.notifyNodeChanged();
+		}
+		return deleted;
+	}
+
+	public notifyNodeChanged(node?: TreeNode): void {
+		this._onDidChangeTreeData.fire(node);
+	}
+
+	public async loadSavedControllers(): Promise<void> {
+		let config = vscode.workspace.getConfiguration(ConfigNamespace);
+		if (config && config.controllers) {
+			let controllers = config.controllers;
+			this.root.clearChildren();
+			for (let c of controllers) {
+				let password = undefined;
+				if (c.rememberPassword) {
+					password = await this.getPassword(c.url, c.username);
+				}
+				this.root.addChild(new ControllerNode({
+					url: c.url,
+					username: c.username,
+					password: password,
+					rememberPassword: c.rememberPassword,
+					parent: this.root,
+					treeChangeHandler: this
+				}));
+			}
+			this.notifyNodeChanged();
+		}
+	}
+
+	public async saveControllers(): Promise<void> {
+		let controllers = this.root.children.map(e => {
+			let controller = e as ControllerNode;
+			return {
+				url: controller.url,
+				username: controller.username,
+				password: controller.password,
+				rememberPassword: !!controller.rememberPassword
+			};
+		});
+
+		let controllersWithoutPassword = controllers.map(e => {
+			return {
+				url: e.url,
+				username: e.username,
+				rememberPassword: e.rememberPassword
+			};
+		});
+
+		try {
+			await vscode.workspace.getConfiguration(ConfigNamespace).update('controllers', controllersWithoutPassword, true);
+		} catch (error) {
+			vscode.window.showErrorMessage(error.message);
+		}
+
+		for (let e of controllers) {
+			await this.savePassword(e.url, e.username, e.password);
+		}
+	}
+
+	private async savePassword(url: string, username: string, password: string): Promise<boolean> {
+		let provider = await this.getCredentialProvider();
+		let id = this.createId(url, username);
+		let result = await provider.saveCredential(id, password);
+		return result;
+	}
+
+	private async getPassword(url: string, username: string): Promise<string> {
+		let provider = await this.getCredentialProvider();
+		let id = this.createId(url, username);
+		let credential = await provider.readCredential(id);
+		return credential ? credential.password : undefined;
+	}
+
+	private async getCredentialProvider(): Promise<azdata.CredentialProvider> {
+		if (!this.credentialProvider) {
+			this.credentialProvider = await azdata.credentials.getProvider(CredentialNamespace);
+		}
+		return this.credentialProvider;
+	}
+
+	private createId(url: string, username: string): string {
+		return `${url}::${username}`;
```

Comment:
We need all Ids to be globally unique (e.g. what if another extension also used a URL::user combo)? Please add the extension id, e.g. `big-data-cluster::${url}::${username}`

Reply:
Or maybe that's what the CredentialNamespace is for? If that does the credential namespacing, I'm happy. I'm referring to:
```ts
			this.credentialProvider = await azdata.credentials.getProvider(CredentialNamespace);
```